### PR TITLE
Implement `array::_array_reconstructor`

### PR DESCRIPTION
**Pull request by whjpji**
Test cases:
- [x] `ArrayReconstructorTest.test_error`
- [x] `ArrayReconstructorTest.test_numbers`
- [x] `ArrayReconstructorTest.test_unicode`
    - [x] `'UTF-16-LE'`
    - [x] `'UTF-16-BE'`
    - [ ] ~`'UTF-32-LE'`~
    - [ ] ~`'UTF-32-BE'`~

UTF32 related work will be done later in a dedicated PR.

whitespace-ignoring review link: https://github.com/RustPython/RustPython/pull/2937/files?w=1

### CODE REVIEW

**Comment by whjpji** [comment]
Hi, @coolreader18, I'm working with the UTF32 encodings/decodings. There are already some enc/dec functions in [`Lib/_pycodecs.py`](https://github.com/RustPython/RustPython/blob/main/Lib/_pycodecs.py).

There is a comment says that it comes from PyPy, but I cannot find the source in PyPy's code base, but only find [this file](https://foss.heptapod.net/pypy/pypy/-/blob/branch/default/pypy/module/_codecs/interp_codecs.py). It is very different from `_pycodecs.py`.

Did you remember where the `_pycodecs.py` come from? I might quickly add the enc/dec for UTF32, and maybe rewrite them in Rust in the next PR.

**Comment by whjpji** [comment]
@youknowone No problem. Let me resolve the conflicting files first.

**Comment by coolreader18** [comment]
One last thing - could you separate this into 2 commits? One with reformatting `array.rs` to be `mod array { ... }`, the other with the changes you made for `_array_reconstructor`? Hopefully that shouldn't be too hard :grimacing:

**Comment by whjpji** [comment]
Oh, sad! The indents in marcos of the first commit is not correct, let me fix it.

**Review by youknowone — CHANGES_REQUESTED** [review]

File: vm/src/stdlib/array.rs [hunk 1]
```diff
@@ -1,1067 +1,1418 @@
...
+            })?,
+            MachineFormatCode::Int64 { signed, big_endian } => chunks.try_for_each(|b| {
+                array.push(chunk_to_obj!(vm, b, i64, u64, signed, big_endian), vm)
+            })?,
+            MachineFormatCode::Utf16 { big_endian } => {
+                let utf16: Vec<_> = chunks.map(|b| chunk_to_obj!(b, u16, big_endian)).collect();
+                let s = String::from_utf16(&utf16).map_err(|_| {
+                    vm.new_unicode_encode_error("items cannot decode as utf16".into())
+                })?;
+                eprintln!("utf16: {:?}", s);
```
Comment (youknowone) [inline vm/src/stdlib/array.rs:1398]: these `eprintln`s look like debug prints probably not inteded to be included in final PR

**Comment by youknowone** [comment]
I left 2 more comments and other parts looks good.

**Review by youknowone — APPROVED** [review]

**Pull request merged**